fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java

```diff
@@ -0,0 +1,132 @@
+
+@Slf4j
+public final class DelinquencyUtils {
+
+    private DelinquencyUtils() {}
+
+    public static CollectionData getOverdueCollectionData(final Loan loan, final LocalDate businessDate) {
+        final MonetaryCurrency loanCurrency = loan.getCurrency();
+        LocalDate overdueSinceDate = null;
+        CollectionData collectionData = CollectionData.template();
+        BigDecimal amountAvailable = BigDecimal.ZERO;
+        BigDecimal outstandingAmount = BigDecimal.ZERO;
+        boolean oldestOverdueInstallment = false;
+        boolean overdueSinceDateWasSet = false;
+        boolean firstNotYetDueInstallment = false;
+
+        log.debug("Loan id {} with {} installments", loan.getId(), loan.getRepaymentScheduleInstallments().size());
+        // Get the oldest overdue installment if exists one
+        for (LoanRepaymentScheduleInstallment installment : loan.getRepaymentScheduleInstallments()) {
+            if (installment.isObligationsMet()) {
+                continue;
+            }
+
+            if (installment.getDueDate().isBefore(businessDate)) {
+                log.debug("Loan Id: {} with installment {} due date {}", loan.getId(), installment.getInstallmentNumber(),
+                        installment.getDueDate());
+                outstandingAmount = outstandingAmount.add(installment.getTotalOutstanding(loanCurrency).getAmount());
+                if (!oldestOverdueInstallment) {
+                    log.debug("Oldest installment {} {}", installment.getInstallmentNumber(), installment.getDueDate());
+                    oldestOverdueInstallment = true;
+                    overdueSinceDate = installment.getDueDate();
+                    overdueSinceDateWasSet = true;
+
+                    amountAvailable = installment.getTotalPaid(loanCurrency).getAmount();
+
+                    for (LoanTransactionToRepaymentScheduleMapping mappngInstallment : installment
+                            .getLoanTransactionToRepaymentScheduleMappings()) {
+                        final LoanTransaction loanTransaction = mappngInstallment.getLoanTransaction();
+                        if (loanTransaction.isChargeback()) {
+                            amountAvailable = amountAvailable.subtract(loanTransaction.getAmount());
+                            if (amountAvailable.compareTo(BigDecimal.ZERO) < 0) {
+                                overdueSinceDate = loanTransaction.getTransactionDate();
+                                break;
+                            }
+                        }
+                    }
+                }
+            } else if (!firstNotYetDueInstallment) {
+                log.debug("Loan Id: {} with installment {} due date {}", loan.getId(), installment.getInstallmentNumber(),
+                        installment.getDueDate());
+                firstNotYetDueInstallment = true;
+                amountAvailable = installment.getTotalPaid(loanCurrency).getAmount().subtract(installment.getTotalPaidInAdvance());
```

Comment:
Are you sure we need to deduct the total paid in advance?

Reply:
Removed